servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SingleAmbWith.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright © 2020 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.concurrent.api.AmbSingles.AmbSubscriber;
+import io.servicetalk.concurrent.api.AmbSingles.State;
+import io.servicetalk.concurrent.internal.SignalOffloader;
+
+import static java.util.Objects.requireNonNull;
+
+final class SingleAmbWith<T> extends AbstractNoHandleSubscribeSingle<T> {
+    private final Single<T> original;
+    private final Single<T> ambWith;
+
+    SingleAmbWith(final Executor executor, final Single<T> original, final Single<T> ambWith) {
+        super(executor);
+        this.original = requireNonNull(original);
+        this.ambWith = requireNonNull(ambWith);
+    }
+
+    @Override
+    void handleSubscribe(final Subscriber<? super T> subscriber, final SignalOffloader signalOffloader,
+                         final AsyncContextMap contextMap, final AsyncContextProvider contextProvider) {
+        State<T> state = new State<>(subscriber);
+        subscriber.onSubscribe(state);
+        AmbSubscriber<T> originalSubscriber = new AmbSubscriber<>(state);
```

Comment:
consider including this inside the `try` catch to avoid letting the exception escape.